### PR TITLE
Feature/refactor auth flow

### PR DESCRIPTION
**Pull request by fisher60**
Completely redesign the auth flow of Abandon Auth to use AA UI and redirects to complete the authentication process.

This PR is a WIP and I am creating for visibility as there will be some very important changes that completely overhaul the auth flow.

### CODE REVIEW

File: abandonauth/dependencies/auth/jwt.py [hunk 1]
```diff
@@ -3,19 +3,23 @@
...
 from abandonauth.settings import settings
 
 # Cache of all valid issued tokens. Tokens should be removed after their first use
 valid_token_cache = set()
 
+IGNORE_AUD_DECODE_OPTIONS = {"verify_aud": False}
 
-def _generate_jwt(user_id: str, long_lived: bool = False) -> str:
+
+def _generate_jwt(user_id: str, application_id_aud: str, long_lived: bool = False) -> str:
```
Comment (GDWR) [inline abandonauth/dependencies/auth/jwt.py:18]: is `application_id_aud` different to `application_id`?

File: abandonauth/dependencies/auth/jwt.py [hunk 2]
```diff
@@ -26,12 +30,21 @@ def _generate_jwt(user_id: str, long_lived: bool = False) -> str:
 
     expiration = datetime.now(timezone.utc) + timedelta(seconds=exp_seconds)
 
+    if long_lived and application_id_aud == settings.ABANDON_AUTH_DEVELOPER_APP_ID:
+        scope = " ".join((ScopeEnum.abandonauth, ScopeEnum.identify))
```
Comment (GDWR) [inline abandonauth/dependencies/auth/jwt.py:34]: we can look at flag enums in the future, this isn't the nicest

File: abandonauth/models/developer_application.py [hunk 3]
```diff
@@ -8,6 +8,12 @@ class DeveloperApplicationDto(BaseModel):
     owner_id: str
 
 
+class DeveloperApplicationWithCallbackUriDto(DeveloperApplicationDto):
+    """Basic data for developer applications as well as the Callback URIs for the app"""
+
+    callback_uris: list[str] | None
```
Comment (GDWR) [inline abandonauth/models/developer_application.py:14]: I'd prefer the no callback uris type to still be a `list[str]` but just empty.

File: abandonauth/routers/developer_application.py [hunk 4]
```diff
@@ -141,14 +150,106 @@ async def change_application_refresh_token(
...
+    summary="Retrieve the given application if it belongs to the currently authenticated user",
+    response_description="General information about the developer application",
+    response_model=DeveloperApplicationWithCallbackUriDto
+)
+async def get_developer_application(
+    application_id: str,
+    token_data: JwtClaimsDataDto = Depends(JWTBearer())
+) -> DeveloperApplicationWithCallbackUriDto:
+    """Get information about the given developer application if the requesting user owns the developer app."""
+    print("RUNNING GET_DEV_APP")
```
Comment (GDWR) [inline abandonauth/routers/developer_application.py:180]: 🤔

**Review by GDWR — APPROVED** [review]
Added some comments, but happy to merge and have future PR fixes

File: abandonauth/routers/index.py [hunk 5]
```diff
@@ -22,6 +25,12 @@ async def identify_user(user_id: str) -> User:
     return user
 
 
+def get_new_token(exchange_token: str, aud: str | None = None) -> JwtDto:
```
[the remark this answers is not in the document]
Comment (fisher60) [reply abandonauth/routers/index.py:28]: This aud is used to decode the existing jwt, a JWT with aud cannot be decoded unless the aud is verified, therefore this is likely to be required depending what the required aud is when decoding the original JWT.

File: abandonauth/dependencies/auth/jwt.py (hunk 1 above)
Comment (fisher60) [reply abandonauth/dependencies/auth/jwt.py:18]: No, these values are identical. I felt it was a bit more clear to specify what the application ID is used for in this case though.

File: abandonauth/models/developer_application.py (hunk 3 above)
Comment (fisher60) [reply abandonauth/models/developer_application.py:14]: Good catch! We were actually returning an empty list here, but the typing on the DTO was incorrect.

File: abandonauth/routers/developer_application.py (hunk 4 above)
Comment (fisher60) [reply abandonauth/routers/developer_application.py:180]: [Fix for this](https://github.com/jayphelps/git-blame-someone-else)

**Comment by fisher60** [comment]
closes #58

**Pull request merged**